README.md
# Memcing

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Memcing is an in-memory key-value caching application with RESTful API.

## Installation

```bash
npm install memcing
```

## Usage

### Interactive mode

```bash
npm start
```
Add an entry (`add hello world`) and see `http://localhost:12080/entries/hello`

### Help

```bash
./memcing -help
```
```
  Options:
    -i                 : Enable interactive mode.
    -listen-http       : Listen HTTP requests. Default; 0.0.0.0:12080
    -load-file         : Load a command file.

    -cache-limit       : Cache size limit in bytes. Default (128MB); 134217728
    -entry-limit       : Entry size limit in bytes. Default (1KB); 1024
    -vacuum-delay      : Delay in seconds for vacuum. Default; 30
    -eviction          : Enable eviction mode.

    -debug             : Enable debug mode.
    -verbose           : Set verbose message level. Default; 1
    -help              : Display help and exit.

  stdin:
    -cmd               : Enable command mode for stdin.
    -csv               : Enable CSV mode for stdin.
    -csv-delimiter     : CSV delimiter (char or `tab`). Default; ,
    -csv-field-key     : Key field index on CSV. Default; 1
    -csv-field-filter  : Include fields on CSV. Default (all); null
                         Example; -csv-field-filter 1,2

  Interactive mode commands:
    get key
    set key value [expire = 0]
    add key value [expire = 0]
    increment key [amount = 1]
    decrement key [amount = 1]
    delete key
    drop
    dump
    stats
    vacuum
    exit
```

### Examples

#### Interactive Mode

```bash
./memcing -load-file ../test/cmds-lf.txt -i -listen-http
```

Reads commands from `cmds-lf.txt` then switch to interactive mode and listen 
ttp requests. The cached data will be available for REPL and RESTful API.

For RESTful API performance tests;

```bash
ab -n 10000 http://localhost:12080/entries/hello
ab -n 10000 -c 100 http://localhost:12080/entries/hello
```

#### Postal Code Service Example

You can easily create a postal code service;

```bash
wget -qO zip-codes.zip http://download.geonames.org/export/zip/US.zip && \
unzip -p zip-codes.zip US.txt | \
cat | \
./memcing -listen-http -csv -csv-delimiter tab -csv-field-key 2 -csv-field-filter 1,3,4,5,6,10,11
```

See `http://localhost:12080/entries/78729`

```JSON
{
  "key":"78729",
  "val":"[\"US\",\"Austin\",\"Texas\",\"TX\",\"Williamson\",\"30.4521\",\"-97.7688\"]",
  "ts":1401417873489,
  "expTS":0
}
```

If you want to include other countries see http://download.geonames.org/export/zip/
and change the part (`US.zip` and `US.txt`) of the command at above.

## RESTful API

*RESTful API is still under development. For the current implementations see [Implementations](#implementations)* 

* Use `application/x-www-form-urlencoded` for request method.
* Use `val` query parameter for value.
* Use `exp` query parameter for expiration.

### PUT

```
/entries/{KEY}
```
PUT method represent the `set` command.

Examples:
```
curl -X PUT -d "val=world" http://localhost:12080/entries/hello
curl -X PUT -d "val=world&exp=10" http://localhost:12080/entries/bye
```

### POST

```
/entries/{KEY}
```
POST method represent the `add` command.

Examples:
```
curl -X POST -d "val=1" http://localhost:12080/entries/counter
```

### GET

```
/entries
/entries/{KEY}
```
GET method represent the `get` command.

Examples:
```
curl http://localhost:12080/entries
curl http://localhost:12080/entries/hello
```

### DELETE

```
/entries
/entries/{KEY}
```
DELETE method represent the `delete` or the `drop` command.

Examples:
```
curl -X DELETE http://localhost:12080/entries/hello
curl -X DELETE http://localhost:12080/entries
```

## Notes

* It is 'originally' developed for code exercise. The requirements were;
  non-persistent, non-durable and supporting eviction.

## Implementations

- [x] Interactive mode
- [ ] RESTful API
  - [x] GET
    - [x] Collection (`/entries`)
    - [x] Element (`/entries/{KEY}`)
  - [ ] PUT
    - [ ] Collection (`/entries`)
    - [x] Element (`/entries/{KEY}`)
  - [ ] POST
    - [ ] Collection (`/entries`)
    - [x] Element (`/entries/{KEY}`)
  - [x] DELETE
    - [x] Collection (`/entries`)
    - [x] Element (`/entries/{KEY}`)

###License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/memcing
[npm-image]: https://badge.fury.io/js/memcing.svg

[travis-url]: https://travis-ci.org/devfacet/memcing
[travis-image]: https://travis-ci.org/devfacet/memcing.svg?branch=master


CHANGELOG.md
# Changelog

## 1.0.0 (2017-03-05)

- Maintenance

## 0.2.10 (2015-08-08)

- Misc. updates

## 0.2.9 (2015-03-15)

- Misc. updates

## 0.2.8 (2014-07-04)

- Misc. updates

## 0.2.7 (2014-06-16)

- Misc. updates

## 0.2.6 (2014-05-29)

- Misc. updates

## 0.2.5 (2014-05-28)

- Add new tests

## 0.2.4 (2014-05-26)

- Add JS linter rules
- Change ignore rules
- Add test for options
- Add test for stdin

## 0.2.3 (2014-05-21)

- Misc. updates

## 0.2.2 (2014-05-12)

- Add init module
- Changes on package.json
- Misc. updates

## 0.2.1 (2014-05-08)

- Change for stdin read buffer size

## 0.2.0 (2014-05-06)

- Add csv support
- Add pipe module
- Add verbose option
- Add auto completion for REPL
- More debugging messages

## 0.1.7 (2014-04-30)

- Add benchmark script for basic ops
- Change -cache-limit (size in bytes)
- Add -entry-limit
- Add -vacuum-delay (remove -vacuum-ival)
- Misc. updates for roadmap (binary arrays)
- Add Dependency Status
- Add appveyor.yml

## 0.1.6 (2014-04-25)

- Add npm badge
- Add Build Status
- Add .travis.yml
- Add test for cache module
- Add test for rest module
- Add addrOf - rest module
- Fix hostname option
- Behavior changes for --listen-http

## 0.1.5 (2014-04-22)

- Changes for coding style
- repl module
- rest module
- Refactoring cache module

## 0.1.4 (2014-04-15)

- drop command
- Add REST method (DELETE - collection)
- Add REST method (DELETE - element)
- Add REST method (POST - element)
- Add REST method (PUT - element)
- Add REST method (GET - collection)

## 0.1.3 (2014-04-12)

- Add REST method (GET - element)
- Add heapsnapshot for tests
- Changes on min eviction limit
- Improvement for vacuuming

## 0.1.2 (2014-04-09)

- Changes for eviction.
- Better vacuuming.

## 0.1.1 (2014-04-08)

- Misc. improvements.

## 0.1.0 (2014-04-07)

- Initial release

index.js
/*
 * Memcing
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var utilex = require('utilex'),
    fs     = require('fs'),
    path   = require('path'),
    init   = require('./lib/init'),
    cache  = require('./lib/cache'),
    pipe   = require('./lib/pipe'),
    rest   = require('./lib/rest'),
    repl   = require('./lib/repl');

var appConfig = {appPath: fs.realpathSync(path.join(__dirname, '..'))},
    appPIPE,  // pipe instance
    appCache, // cache instance
    appREST,  // rest instance
    appREPL;  // repl instance

// Init the app
init({appConfig: appConfig});

// Check for the help command
if(!appConfig.loadFile && !appConfig.listen && !appConfig.iactive && process.stdin.isTTY === true) cmdHelp();

// Create the instances
appCache = cache(appConfig.cache);
appPIPE  = pipe(appConfig.pipe, appCache);
appREST  = rest(appConfig.rest, appCache);
appREPL  = repl(appConfig.repl, appCache);

appCache.loadFile(appConfig.loadFile)
.then(appPIPE.start)
.then(appREST.listen)
.then(appREPL.start)
.then(function(res) {
  if(res) utilex.conLog(res);
})
.catch(function(err) {
  utilex.conLog(err);
  process.exit(0);
})
.done(function() {
  if(!appConfig.listen && !appConfig.iactive) {
    process.exit(0); // Nothing to do
  }
});

// Displays help and exit.
function cmdHelp() {

  console.log("Usage: node memching.js [OPTION]...\n");

  console.log("Memcing is an in-memory key-value caching application.\n");

  console.log("  Options:");
  console.log("    -i                 : Enable interactive mode.");
  console.log("    -listen-http       : Listen HTTP requests. Default; 0.0.0.0:12080");
  console.log("    -load-file         : Load a command file.\n");

  console.log("    -cache-limit       : Cache size limit in bytes. Default (128MB); 134217728");
  console.log("    -entry-limit       : Entry size limit in bytes. Default (1KB); 1024");
  console.log("    -vacuum-delay      : Delay in seconds for vacuum. Default; 30");
  console.log("    -eviction          : Enable eviction mode.\n");

  console.log("    -debug             : Enable debug mode.");
  console.log("    -verbose           : Set verbose message level. Default; 1");
  console.log("    -help              : Display help and exit.\n");

  console.log("  stdin:");
  console.log("    -cmd               : Enable command mode for stdin.");
  console.log("    -csv               : Enable CSV mode for stdin.");
  console.log("    -csv-delimiter     : CSV delimiter (char or `tab`). Default; ,");
  console.log("    -csv-field-key     : Key field index on CSV. Default; 1");
  console.log("    -csv-field-filter  : Include fields on CSV. Default (all); null");
  console.log("                         Example; -csv-field-filter 1,2\n");

  console.log("  Interactive mode commands:");
  console.log("    get key");
  console.log("    set key value [expire = 0]");
  console.log("    add key value [expire = 0]");
  console.log("    increment key [amount = 1]");
  console.log("    decrement key [amount = 1]");
  console.log("    delete key");
  console.log("    drop");
  console.log("    dump");
  console.log("    stats");
  console.log("    vacuum");
  console.log("    exit\n");

  console.log("  Examples:");
  console.log("    node memcing.js -i");
  console.log("    node memcing.js -i -load-file /path/file -listen-http");
  console.log("    echo hello world | node memcing.js -listen-http");
  console.log("    cat file.csv | node memcing.js -listen-http -csv");
  console.log("    cat cmds.txt | node memcing.js -listen-http -cmd\n");

  process.exit(0);
}